docs/book/src/clusterctl/contract.md

```diff
@@ -0,0 +1,207 @@
+
+<aside class="note warning">
+
+<h1>Warning</h1>
+
+If the generated component YAML does't contain a Namespace object, the user will be required to provide one to `clusterctl init` 
+using the `--target-namespace` flag.
+
+In case there is more than one Namespace object in the components YAML, `clusterctl` will generate an error and abort
+the provider installation.
+
+</aside>
+
+#### Controllers & Watching namespace
+
+Each provider is expected to deploy controllers using a Deployment.
+
+While defining the Deployment Spec, the container that executes the controller binary MUST be called `manager`.
+ 
+The manager MUST support a `--namespace` flag for specifying the namespace where the controller
+will look for objects to reconcile.
+
+#### Variables
+
+The components YAML can contain environment variables matching the regexp `\${\s*([A-Z0-9_]+)\s*}`; it is highly
+recommended to prefix the variable name with the provider name e.g. `${ AWS_CREDENTIALS }`
+
+Additionally, each provider should create user facing documentation with the list of required variables and with all the additional
+notes that are required to assist the user in defining the value for each variable.
+ 
+### Workload cluster templates
+
+An infrastructure provider could publish a **cluster templates** file to be used by `clusterctl config cluster`. 
+This is single YAML with _all_ the objects required to create a new workload cluster. 
+
+The following rules apply:
+
+#### Naming conventions
+
+Cluster templates MUST be stored in the same folder as the component YAML and follow this naming convention:
+1. The default cluster template should be named `config-{bootstrap}.yaml`. e.g `config-kubeadm.yaml`
+2. Additional cluster template should be named `config-{flavor}-{bootstrap}.yaml`. e.g `config-production-kubeadm.yaml`
+
+`{bootstrap}` is the name of the bootstrap provider used in the template; `{flavor}` is the name the user can pass to the
+`clusterctl config cluster --flavor` flag to identify the specific template to use.
+ 
+Each provider SHOULD create user facing documentation with the list of available cluster templates.
+
+#### Target namespace
+
+The cluster template YAML MUST assume the target namespace already exists.
+
+All the objects in the cluster template YAML MUST be deployed in the same namespace. 
+
+#### Variables
+
+The cluster templates YAML can also contain environment variables (as can the components YAML).
+
+Additionally, each provider should create user facing documentation with the list of required variables and with all the additional
+notes that are required to assist the user in defining the value for each variable.
+
+##### Common variables
+
+The `clusterctl config cluster` command allows user to set a small set of common variables via CLI flags or command arguments.
+
+Templates writers should use the common variables to ensure consistency across providers and a simpler user experience
```

Comment:
Probably stronger than 'should'? Given the current implementation, the use of these variable names for the purpose they are intended is mandantory.

Also the absence of some these variables can lead to surprises in the UX. Say, if a template is opinionated about the number of control plane nodes it should create (e.g. a 'dev' template that hard codes that number to 1 to minimize resource usage), a user attempting to use said template while specifying 
```... --controlplane-machine-count 3 --worker-machine-count 5 
```
would see only part of his intent satisfied in an otherwise fully functional configuration without warning

Reply:
This is correct.
The authors of templates should provide documentation, and in case the template is opinionated, also the underlying assumptions (e.g. 1 worker to minimize resources) should be documented.

However, if this became a problem, we should always change this into a minimal set of required variables, but TBH, I fear this is too much restrictive ATM